### PR TITLE
AppUser export support

### PR DESCRIPTION
**Pull request by solocommand**
Adds support for exporting app user data

Screen Shot 2020-01-30 at 5 01 31 PM
Screen Shot 2020-01-30 at 5 02 41 PM
Screen Shot 2020-01-30 at 5 01 57 PM
Screen Shot 2020-01-30 at 5 01 42 PM

### CODE REVIEW

File: services/export/src/actions/user/export-for-app.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  email,
+  applicationId,
+} = {}) => {
+  if (!email) throw createRequiredParamError('email');
+  if (!applicationId) throw createRequiredParamError('applicationId');
+  if (!Array.isArray(fields)) throw createRequiredParamError('fields');
+  const projection = fields.reduce((obj, k) => ({ ...obj, [k]: 1 }), {});
+
+  try {
+    const users = await AppUser.find({ applicationId }, projection);
```
Comment (zarathustra323) [inline services/export/src/actions/user/export-for-app.js]: Use application service

File: services/export/src/actions/user/export-for-app.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    const contents = await parse(users, { fields });
+    const filename = `export-${applicationId}-${Date.now()}.csv`;
```
Comment (zarathustra323) [inline services/export/src/actions/user/export-for-app.js]: `app-user-export-${applicationId}-${Date.now()}.csv`

File: services/export/src/s3.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+module.exports = {
+  client,
+  upload: ({ filename, contents }) => new Promise((resolve, reject) => client.putObject({
+    ACL: 'public-read',
+    Body: contents,
+    Bucket: AWS_S3_BUCKET_NAME,
+    ContentDisposition: `attachment; filename="${filename}"`,
+    ContentType: 'text/plain; charset=utf8',
+    Expires: Math.floor(Date.now().valueOf() / 1000),
```
Comment (zarathustra323) [inline services/export/src/s3.js]: Times 30 days

File: services/manage/app/components/button/export.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+const mutation = gql`
+  mutation RequestExportAppUsers {
+    exportAppUsers
+  }
+`;
+
+export default Component.extend(ActionMixin, ObjectQueryManager, {
+  errorNotifier: inject(),
+  context: inject(),
```
Comment (zarathustra323) [inline services/manage/app/components/button/export.js]: I believe this can be mixed in.

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: services/manage/app/components/button/export.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import gql from 'graphql-tag';
+
+const mutation = gql`
+  mutation RequestExportAppUsers {
+    exportAppUsers
+  }
+`;
+
+export default Component.extend(ActionMixin, ObjectQueryManager, {
+  errorNotifier: inject(),
```
Comment (zarathustra323) [inline services/manage/app/components/button/export.js]: Move the action logic outside of the component

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Comment by solocommand** [comment]
@zarathustra323 all changes implemented

File: services/export/package.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  "dependencies": {
+    "@base-cms/env": "^0.9.25",
+    "@base-cms/id-me-service-clients": "^0.7.0",
+    "@base-cms/micro": "^0.9.28",
+    "@base-cms/object-path": "^1.9.0",
+    "@newrelic/native-metrics": "^4.1.0",
+    "aws-sdk": "^2.610.0",
+    "json2csv": "^4.5.4",
+    "micro": "^9.3.4",
+    "mongoose": "^5.5.11",
```
Comment (zarathustra323) [inline services/export/package.json]: Remove mongoose.

File: services/export/src/utils/stream-results.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
+const { getAsArray, getAsObject } = require('@base-cms/object-path');
+
+/**
+ * Pipes results to the passed input stream
+ * @param client      - The service client
```
Comment (zarathustra323) [inline services/export/src/utils/stream-results.js]:
```
@param {object} args The function args
@param {object} args.client
@param {string} args.action The service action
```

File: services/export/src/utils/stream-results.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  } = args;
+  const pagination = getAsObject(params, 'pagination');
+  const data = await client.request(action, {
+    ...params,
+    pagination: {
+      ...pagination,
+      limit,
+    },
+  });
+  const nodes = getAsArray(data, 'edges').map(({ node }) => node);
```
Comment (zarathustra323) [inline services/export/src/utils/stream-results.js]: `const nodes = getAsArray(data, 'edges').map(edge => edge.node);`

File: services/export/src/utils/stream-results.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
+const { getAsArray, getAsObject } = require('@base-cms/object-path');
+
+/**
+ * Pipes results to the passed input stream
+ * @param client      - The service client
+ * @param action      - The service action
+ * @param params      - The service params
+ * @param limit       - The number of results that should be returned with each page
+ * @param stream      - The input stream to send results to
```
Comment (zarathustra323) [inline services/export/src/utils/stream-results.js]: `The JSON2CSV stream... yadda yadda`

File: services/export/src/utils/stream-results.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      params: {
+        ...params,
+        pagination: { ...pagination, after: endCursor },
+      },
+    });
+  }
+  stream.push(null);
+};
+
+module.exports = async ({
```
Comment (zarathustra323) [inline services/export/src/utils/stream-results.js]: `module.exports = executor`

File: services/export/src/utils/stream-results.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * @param params      - The service params
+ * @param limit       - The number of results that should be returned with each page
+ * @param stream      - The input stream to send results to
+ */
+const executor = async (args) => {
+  const {
+    client,
+    action,
+    params,
+    limit,
```
Comment (zarathustra323) [inline services/export/src/utils/stream-results.js]: `limit = 500`

File: services/manage/app/templates/manage/orgs/view/apps/view/users.hbs [hunk 12]
```diff
@@ -45,4 +45,10 @@
   <Button::Create @routeName="manage.orgs.view.apps.view.users.create" @title="Create New User" />
 </aside>
 
+{{#if (gt this.model.totalCount 0)}}
+  <aside class="footer-controls right">
+    <Button::Export @title="Export Users" @action={{action "export"}} />
```
Comment (zarathustra323) [inline services/manage/app/templates/manage/orgs/view/apps/view/users.hbs]: `@click={{action "export"}}`

File: services/manage/app/components/button/export.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import Component from '@ember/component';
+
+export default Component.extend({
+  tagName: 'button',
+  classNames: 'btn btn-xl btn-info btn-circle btn-create',
+  icon: 'download',
+  title: 'Export',
+  action: null,
+
+  async click() {
```
Comment (zarathustra323) [inline services/manage/app/components/button/export.js]: Remove this function

File: services/export/src/actions/user/export-for-app.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    const filename = `app-user-export-${applicationId}-${Date.now()}.csv`;
+    await upload({ filename, contents });
+
+    const url = `https://${AWS_S3_BUCKET_NAME}.s3.amazonaws.com/${filename}`;
+    await sendSuccess({ email, url });
+    return 'ok';
+  } catch (e) {
+    newrelic.noticeError(e);
+    await sendFailure({ email, error: e });
```
Comment (zarathustra323) [inline services/export/src/actions/user/export-for-app.js]: `sendFailure({ email, error: e }).catch(newrelic.noticeError)`

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Review by zarathustra323 — APPROVED** [review]

**Pull request merged**